tools/run_tests/python_utils/bazel_report_helper.py

```diff
@@ -0,0 +1,252 @@
+        (report_suite_name, package_name),
+        '<testcase name="%s">' % testcase_name,
+        test_output_tag,
+        '</testcase>'
+        '</testsuite>',
+        '</testsuites>',
+    ]
+    return '\n'.join(lines)
+
+
+def _create_bazel_wrapper(report_path, report_suite_name, invocation_id,
+                          upload_results):
+    """Create a "bazel wrapper" script that will execute bazel with extra settings and postprocessing."""
+
+    os.makedirs(report_path, exist_ok=True)
+
+    bazel_wrapper_filename = os.path.join(report_path, 'bazel_wrapper')
+    bazel_wrapper_bat_filename = bazel_wrapper_filename + '.bat'
+    bazel_rc_filename = os.path.join(report_path, 'bazel_rc')
+
+    # put xml reports in a separate directory if requested by GRPC_TEST_REPORT_BASE_DIR
+    report_base_dir = os.getenv('GRPC_TEST_REPORT_BASE_DIR', None)
+    xml_report_path = os.path.abspath(
+        os.path.join(report_base_dir, report_path
+                    ) if report_base_dir else report_path)
+    os.makedirs(xml_report_path, exist_ok=True)
+
+    failing_report_filename = os.path.join(xml_report_path, 'sponge_log.xml')
+    success_report_filename = os.path.join(xml_report_path,
+                                           'success_log_to_rename.xml')
+
+    if _platform_string() == 'windows':
+        workspace_status_command = 'tools/remote_build/workspace_status_kokoro.bat'
+    else:
+        workspace_status_command = 'tools/remote_build/workspace_status_kokoro.sh'
+
+    # generate RC file with the bazel flags we want to use apply.
+    # Using an RC file solves problems with flag ordering in the wrapper.
+    # (e.g. some flags need to come after the build/test command)
+    with open(bazel_rc_filename, 'w') as f:
+        f.write('build --invocation_id="%s"\n' % invocation_id)
+        f.write('build --workspace_status_command="%s"\n' %
+                workspace_status_command)
+
+    # generate "failing" and "success" report
+    # the "failing" is named as "sponge_log.xml", which is the name picked up by sponge/resultstore
+    # so the failing report will be used by default (unless we later replace the report with
+    # one that says "success"). That way if something goes wrong before bazel is run,
+    # there will at least be a "failing" target that indicates that (we really don't want silent failures).
+    with open(failing_report_filename, 'w') as f:
+        f.write(
+            _generate_junit_report_string(report_suite_name,
+                                          invocation_id,
+                                          success=False))
+    with open(success_report_filename, 'w') as f:
+        f.write(
+            _generate_junit_report_string(report_suite_name,
+                                          invocation_id,
+                                          success=True))
+
+    # generate the bazel wrapper for linux/macos
+    with open(bazel_wrapper_filename, 'w') as f:
```

Comment:
Have we considered the alternative of creating one or more Bash scripts accepting different configs (than trying to generate a bash script)? Current version works, but we might need to maintain the generation script for years. Any effort making it easier to maintain might be quite rewarding down the line.

Reply:
I considered having wrapper scripts at first, but I wasn't happy with the result. What I ran into:
- we need the logic for both a .sh wrapper and .bat wrapper. if there's a common python script, most of the "work" (e.g. generate the UUID, generate the rc file, generate the report etc.) can be done by the python script (with good comments and readability), while we just bake the hardcoded results into the generated wrapper. In bash/bat, even "trivial" task quickly become a pain  and the resulting "code" has bad readability and also is duplicated.
- in python, coming up with future tweaks is much simpler than in .sh/.bat
- with pre-existing bash scripts, passing the configuration to the wrapper script isn't easy. Since we need the wrapper to forward all the command line arguments to bazel, we are left with setting environment variables to pass values like the invocation ID to the script, at which point simply baking the arguments into generated code just comes off as easier, especially since we're already generating other files (e.g. the bazelrc and the xml report).

I've given this quite a bit of thought and the current version is what I believe makes the most sense. I'm definitely open to improvements in the future, but I think it's going to be hard to come up with major simplifications. But yeah, often times simplifying stuff improves our ability to see even more possible improvements, so I'll be looking whether we can improve this even further (as followup work).

Reply:
Good point about some logic is much easier to implement in Python than bash or batch script. If the options are using Python to wrap shell script vs. using shell script to wrap Python commands, the former is more manageable and clear (with the comments and abstractions).